src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -0,0 +1,231 @@
+import is.hail.variant._
+import org.apache.hadoop.conf.Configuration
+import org.apache.spark.rdd.RDD
+
+import scala.language.implicitConversions
+import scala.io.Source
+
+object LoadMatrix {
+
+  def warnDuplicates(ids: Array[String]) {
+    val duplicates = ids.counter().filter(_._2 > 1)
+    if (duplicates.nonEmpty) {
+      warn(s"Found ${duplicates.size} duplicate ${plural(duplicates.size, "sample ID")}:\n  @1",
+        duplicates.toArray.sortBy(-_._2).map { case (id, count) => s"""($count) "$id"""" }.truncatable("\n  "))
+    }
+  }
+
+  // this assumes that col IDs are in last line of header.
+  /// FIXME: Is the toString.split call too slow?
+  def parseHeader(lines: Array[String], sep: String = "\t"): Array[String] = {
+    lines.last.toString.split(sep)
+  }
+
+  def getHeaderLines[T](hConf: Configuration, file: String, nLines: Int = 1): Array[String] = hConf.readFile(file) { s =>
+    Source.fromInputStream(s)
+      .getLines()
+      .take(nLines)
+      .toArray
+  }
+
+  def apply(hc: HailContext,
+            file1: String,
+            files: Array[String],
+            nPartitions: Option[Int] = None,
+            dropSamples: Boolean = false,
+            sep: String = "\t"):
+  VariantSampleMatrix[Annotation, Annotation, Annotation] = {
+    val sc = hc.sc
+    val hConf = hc.hadoopConf
+
+    val header1 = parseHeader(getHeaderLines(hConf, file1), sep)
+    val header1Bc = sc.broadcast(header1)
+
+    val sampleIds: Array[String] =
+      if (dropSamples)
+        Array.empty
+      else
+        header1
+
+    val nSamples = sampleIds.length
+
+    LoadMatrix.warnDuplicates(sampleIds)
+
+    val lines = sc.textFilesLines(files, nPartitions.getOrElse(sc.defaultMinPartitions))
+
+    val fileByPartition = lines.partitions.map(p => partitionPath(p))
+    val firstPartitions = fileByPartition.zipWithIndex.filter((name) => name._2 == 0 || fileByPartition(name._2-1) != name._1).map((name) => name._2)
+
+    val matrixType = MatrixType(VSMMetadata(
+      sSignature = TString,
+      vSignature = TString,
+      genotypeSignature = TInt64
+    ))
+
+    val keyType = matrixType.kType
+    val rowKeys: RDD[RegionValue] = lines.mapPartitionsWithIndex { (i,it) =>
+
+      if (firstPartitions.contains(i)) {
+        val hd1 = header1Bc.value
+        val hd = it.next().value.split(sep)
+        if (!hd1.sameElements(hd)) {
+          hd1.zipAll(hd, None, None).zipWithIndex.foreach { case ((s1, s2), j) =>
+            if (s1 != s2) {
+              fatal(
+                s"""invalid sample ids: expected sample ids to be identical for all inputs. Found different sample ids at position $j.
+                   |    ${files(0)}: $s1
+                   |    ${fileByPartition(i)}: $s2""".
+                  stripMargin)
+            }
+          }
+        }
+      }
+
+      val region = MemoryBuffer()
+      val rvb = new RegionValueBuilder(region)
+      val rv = RegionValue(region)
+
+      new Iterator[RegionValue] {
```

Comment:
So all this complicated iterator business in LoadVCF was to handle some complex filtering conditions that involved partially parsing the input.  I don't think you need any of that here and can just `it.map { ... }`.